### PR TITLE
Temporary: do not merge

### PR DESCRIPTION
**Pull request by arjunguha**

### CODE REVIEW

File: js/src/main/scala/FunctionLib.scala [hunk 1]
```diff
@@ -116,7 +116,6 @@ object FunctionLib {
       case VConst(CNull()) => VConst(CStr("object"))
       case VConst(CRegex(_)) => VConst(CStr("regex"))
       case VBuiltinFunction(_) => VConst(CStr("function"))
-      case VJSObj(_) => VConst(CStr("object"))
```
Comment (arjunguha) [inline js/src/main/scala/FunctionLib.scala]: Notice that this case was impossible at runtime and is now impossible due to the heap-object / value distinction.

File: js/src/main/scala/Syntax.scala [hunk 2]
```diff
@@ -126,6 +131,19 @@ object Syntax extends SyntaxLike {
     if (hi == lo) hi else VFacet(prin, pos(hi), neg(lo))
   }
 
+  def hfacet(prin: Label, hi: HeapValue, lo: HeapValue): HeapValue = {
```
Comment (rachitnigam) [inline js/src/main/scala/Syntax.scala]: This function is to create a new facet?

File: js/src/main/scala/FunctionLib.scala [hunk 3]
```diff
@@ -135,7 +134,6 @@ object FunctionLib {
   //String functions
   def length(args: List[Value]): Value = args match {
     case VBot :: _ => VBot
-    case VArray(items, vobj) :: Nil => VConst(CNum(items.length))
```
Comment (arjunguha) [inline js/src/main/scala/FunctionLib.scala]: primitive functions are applied to values. VArrays are not values. (References to VArrays are values.) Therefore, the length primitive could never be applied to VArray.

File: js/src/main/scala/Syntax.scala (hunk 2 above)
Comment (arjunguha) [inline js/src/main/scala/Syntax.scala]: Yes, but a heap-facet.

**Pull request closed without merging**